### ISSUE TITLE
Independently computing SAM gradient on multiple accelerators

### ISSUE THREAD

**Issue by evanatyourservice** [davda54/sam#18]
Hello,

Thank you for this great implementation. I have a usage question having to do with multi-GPU training.

There is this line from the paper on page 5:

"To compute the SAM update when parallelizing across multiple accelerators, we divide each data batch
evenly among the accelerators, independently compute the SAM gradient on each accelerator, and
average the resulting sub-batch SAM gradients to obtain the final SAM update."

Do you think doing the first SAM step in your implementation on each accelerator, then averaging the gradients from all the accelerators before the second step would accomplish this?

The code would be:
```
for examples, labels in loader:
    loss = loss_fn(network(examples), labels)
    loss.backward()
    optimizer.first_step(zero_grad=True)
    loss_fn(network(examples), targets).backward()
    reduce_gradients_from_all_accelerators()
    optimizer.second_step(zero_grad=True)
```

I believe not reducing the gradients between the accelerators before the first step would accomplish what they describe in the paper, but I'm looking for a second opinion. Any help is much appreciated.

**Comment by evanatyourservice** [comment]
I just did some testing and this seems to work and be the same way they do it in the paper.

**Comment by evanatyourservice** [comment]
So I haven't done an exact side-by-side study yet, but it seems to improve things. I looked closely at the authors' jax code and yours and I'm sure it's the exact same as the way they do it, so that's good. I'm pretty busy so probably won't get around to a pull request, but yeah you could add this to the readme, might be useful. Cheers! I'll get around to doing an exact side-by-side test and report back

**Comment by guozhiyao** [comment]
I use DDP to parallelize the model training on multiple machines. It seems that DDP will automatically average the gradients on different GPUs during backpropagation.

**Comment by davda54** [comment]
You can explicitly stop averaging across GPUs by calling:

```python
with model.no_sync():
    loss.backward()
```

I hope that helps in your case :)

**Comment by guozhiyao** [comment]
> You can explicitly stop averaging across GPUs by calling:
> 
> ```python
> with model.no_sync():
>     loss.backward()
> ```
> 
> I hope that helps in your case :)

Thanks a lot. But I  mean the DDP can average the gradients automatically, so we don't need to do the `reduce_gradients_from_all_accelerators()`. Is that right?

**Comment by evanatyourservice** [comment]
I was using TPUs when I wrote that example

**Comment by yzlnew** [comment]
> You can explicitly stop averaging across GPUs by calling:
> 
> ```python
> with model.no_sync():
>     loss.backward()
> ```
> 
> I hope that helps in your case :)

Should wrap the whole loss calculation to the context. Otherwise it won't work, afaik.

```python
with model.no_sync():
    loss = criterion(model(inputs), targets)
    loss.backward()
```
